deepspeed/runtime/zero/stage3.py

```diff
@@ -1100,7 +1082,13 @@ def __reduce_and_partition_ipg_grads(self, safe_mode: bool = False) -> None:
             if safe_mode:
                 assert_ints_same_as_other_ranks([p.ds_id for p in self.params_in_ipg_bucket])
 
-            grad_partitions = self.__avg_scatter_grads(self.params_in_ipg_bucket)
+            if self.contiguous_gradients and not self.reduce_scatter:
+                grad_bucket = self.__ipg_bucket_flat_buffer.narrow(0, 0, self.elements_in_ipg_bucket)
+                grad_partitions = self.__avg_scatter_contiguous_grads(grad_bucket)
```

Comment:
Please explain more why sorting `self.params._in_ipg_bucket` is not needed and why `self.__ipg_bucket_flat_buffer` is expected to in correct order on all ranks?

Reply:
In __add_grad_to_ipg_bucket line 1055 the current gradient is added to the back of the buffer. The gradients are added in the order the backward hooks are called. I assume backward hooks order is the same in all ranks. Is there a scenario where this is incorrect?
If there is such case isn't it possible that the ranks will try to reduce completly different parameters and the check in line 1074 (new code) will fail?

Reply:
For ZeRo stage 3, I think it is fair to assume that the order of backward hooks and thus gradient creation is identical across the ranks.